packages/@sanity/form-builder/src/inputs/ReferenceInput/ReferenceInput.tsx

```diff
@@ -1,336 +1,311 @@
-        if (typeof filter === 'function') {
-          const parentPath = getValuePath().slice(0, -1)
-          const parent = get(document, parentPath) as Record<string, unknown>
-          return filter({document, parentPath, parent})
-        }
+  const weakIs = value && value._weak ? 'weak' : 'strong'
+  const weakShouldBe = type.weak === true ? 'weak' : 'strong'
+  const isMissing = value?._ref && !preview.isLoading && preview.snapshot === null
 
-        return {filter, params}
-      }
-      search = async (query: string) => {
-        this.setState({
-          isFetching: true,
-        })
+  const hasRef = value && value._ref
 
-        const {type, onSearch} = this.props
-        const options = await this.resolveUserDefinedFilter()
+  const handleFixStrengthMismatch = useCallback(() => {
+    onChange(PatchEvent.from(type.weak === true ? set(true, ['_weak']) : unset(['_weak'])))
+  }, [onChange, type])
 
-        this.subscriptions.replace(
-          'search',
-          onSearch(query, type, options).subscribe({
-            next: (items: Array<SearchHit>) => {
-              this.setState((prev) => {
-                const updatedCache = items.reduce((cache, item) => {
-                  cache[item._id] = item
-                  return cache
-                }, Object.assign({}, prev.refCache))
+  const {push} = useToast()
 
-                return {
-                  hits: items,
-                  isFetching: false,
-                  refCache: updatedCache,
-                }
-              })
-            },
-            error: (err: SearchError) => {
-              const isQueryError = err.details && err.details.type === 'queryParseError'
-              if (!isQueryError || !options.filter) {
-                throw err
-              }
+  const [searchState, setSearchState] = useState<SearchState>(INITIAL_SEARCH_STATE)
 
-              err.message = 'Invalid reference filter, please check `filter`!'
-              throw err
-            },
-          })
-        )
-      }
-      renderHit = (item: SearchHit) => {
-        const type = this.getMemberTypeFor(item._type)
-        return <Preview type={type} value={item} layout="default" />
-      }
-      renderOpenItemElement = () => {
-        const {value} = this.props
-        const {isMissing, previewSnapshot} = this.state
-        if (!value || !value._ref || isMissing) {
-          return null
-        }
-        return (
-          <IntentLink
-            title={previewSnapshot && `Open ${previewSnapshot.title}`}
-            intent="edit"
-            params={{id: value._ref, type: previewSnapshot ? previewSnapshot._type : undefined}}
-            className={styles.referenceLink}
-          >
-            <LinkIcon />
-          </IntentLink>
+  const errors = useMemo(() => markers.filter(isValidationErrorMarker), [markers])
+
+  const handleFocus = useCallback(() => onFocus?.(['_ref']), [onFocus])
```

Comment:
This is a bit hard to read IMO. I’d prefer `if (onFocus) onFocus(['_ref'])` personally